tools/generate-examples.rb

```diff
@@ -0,0 +1,409 @@
+    method = Regexp.last_match(1)
+
+    FileUtils.cp(bgv_file, "#{method}.bgv")
+
+    system("gzip", "-f", "#{method}.bgv")
+    FileUtils.cp("#{method}.bgv.gz", "#{__dir__}/../examples/#{graalvm.name}/#{language}")
+
+    if reference_graalvm?(graalvm)
+      Dir.chdir("#{__dir__}/../examples") do
+        FileUtils.ln_sf(
+          "../#{graalvm.name}/#{language}/#{method}.bgv.gz",
+          "#{language}/#{method}.bgv.gz",
+        )
+      end
+    end
+  end
+end
+
+def log(graalvm, message, io = $stdout)
+  io.puts("#{graalvm.name}: #{message}")
+end
+
+Dir.chdir("tools") do
+  GRAAL_VMS.each do |graalvm|
+    graalvm.install
+
+    ###########################
+    ##### Java: Fibonacci #####
+    ###########################
+
+    graalvm.java_context do |bin|
+      log(graalvm, "Running Fibonacci in Java.")
+
+      FileUtils.cp("#{__dir__}/../examples/Fib.java", ".")
+      system("#{bin}/javac", "Fib.java")
+
+      run_java(bin, 1, "-XX:CompileOnly=Fib::fib", "Fib", "14") do |pipe|
+        loop do
+          line = pipe.gets
+          break if line =~ /4\s+Fib::fib/
+        end
+      end
+
+      bgv = Dir.glob("graal_dumps/**/*.bgv")
+      raise unless bgv.size == 1
+
+      FileUtils.cp(bgv.first, "fib-java.bgv")
+      system("gzip", "-f", "fib-java.bgv")
+      FileUtils.cp("fib-java.bgv.gz", "#{__dir__}/../examples/#{graalvm.name}")
+
+      if reference_graalvm?(graalvm)
+        FileUtils.ln_sf("#{__dir__}/../examples/#{graalvm.name}/fib-java.bgv.gz", "#{__dir__}/../examples")
+      end
+    end
+
+    ###########################
+    ##### Java: Examples  #####
+    ###########################
+
+    graalvm.java_context do |bin|
+      log(graalvm, "Running Java examples.")
+
+      FileUtils.cp("#{__dir__}/../examples/java/JavaExamples.java", ".")
+      system("#{bin}/javac", "JavaExamples.java")
+
+      run_java(bin, 3, "-XX:-UseOnStackReplacement", "-XX:CompileCommand=dontinline,*::*", "JavaExamples") do |pipe|
+        loop do
+          line = pipe.gets if select([pipe], nil, nil, 5)
+          break if line.nil? || line =~ /4\s+Fib::fib/
+        end
+      end
+
+      process_examples(graalvm, "java", "JavaExamples.")
+    end
+
+    #################################
+    ##### JavaScript: Fibonacci #####
+    #################################
+    graalvm.js_context do |bin|
+      log(graalvm, "Running Fibonacci in JavaScript.")
+
+      run_js(bin, "#{__dir__}/../examples/fib.js", "14") do |pipe|
+        loop do
+          line = pipe.gets
+          break if line =~ /opt done\s+(?:id=\d+\s+)?fib/
+        end
+      end
+
+      bgv = Dir.glob('graal_dumps/**/*_fib\\].bgv')
+      bgv_ast = Dir.glob('graal_dumps/**/*\\[fib\\].bgv')
+      # At some point Graal stopped using separate AST files and instead inserted the AST graphs as phases into
+      # the primary graph.
+      has_separate_ast_file = bgv_ast.any?
+
+      raise unless bgv.size == 1
+      raise if bgv_ast.size > 1
```

Comment:
Is there a good error message we could include here?  Like a pointer as to what file to go look at if these sizes aren't right?  Or would it be more likely to be a failure in a command?

Reply:
I really don't know. It's a sanity check so the code that follows makes sense. I can't think of a reason why there would ever be more than one AST file unless Graal changes something unexpectedly. If that's the case, all we can really do is look at the error and reevaluate what assumption no longer holds.